scipy/stats/stats.py

```diff
@@ -4275,6 +4276,164 @@ def mannwhitneyu(x, y, use_continuity=True):
+    p-value set alternative to 'greater' or 'less' (default is 'two-sided').
+
+    For Mann-Whitney-Wilcoxon tests, the reported U statistic is the U used to
+    test the hypothesis. The u1 and u2 statistics are returned as well,
+    corresponding to x and y, respectively.
+
+
+    .. versionadded:: 0.17.0
+
+    References
+    ----------
+    .. [1] H.B. Mann and D.R. Whitney, "On a test of whether one of two random
+           variables is stochastically larger than the other", The Annals of
+           Mathematical Statistics, Vol. 18, pp. 50-60, 1947.
+    .. [2] http://en.wikipedia.org/wiki/Mann-Whitney_U_test
+    """
+    x = asarray(x)
+    y = asarray(y)
+    if len(np.shape(x)) > 1 or len(np.shape(y)) > 1:
+        raise ValueError('Expected 1-d arrays.')
+    n1 = len(x)
+    n2 = len(y)
+    if exact == 'auto':
+        exact = ((n1 < 10 or n2 < 10) and n1 + n2 < 100000
+                 and -np.log(n1 + n2 + 1) - special.betaln(n1 + 1, n2 + 1) < np.log(100000))
+    ranked = rankdata(np.concatenate((x,y)))
+    rankx = ranked[0:n1]       # get the x-ranks
+    T = tiecorrect(ranked)
+    if T == 0:
+        raise ValueError('All numbers are identical')
+    if alternative not in ('two-sided', 'less', 'greater'):
+        raise AttributeError("Alternative should be one of: "
+                             "'two-sided', 'less', or 'greater'")
+    if exact:
+        a = list(range(n1, n1+n2))
+        u = [0]
+        while sum(a) != sum(range(n2)):   # When in leftmost position, a == list(range(n2))
+            # Do the shift operation
+            i = 0
+            while a[i] == i:
+                i += 1
+            a[i] -= 1
+            j = i - 1
+            while j >= 0:
+                a[j] = a[i] - (i - j)
+                j -= 1
+            # count(a < a1) = U2
+            u1 = 0
+            for i, x in enumerate(a):
+                u1 += n1 - x + i
+            u2 = n1*n2 - u1
+            # store min U value to array
+            if alternative == 'two-sided':
+                u.append(min(u1, u2))
+            else:
+                u.append(u1)
+        u1 = 0
+        u = np.array(u)
+        # for i, x in enumerate(sorted(rankx)):
+        #     u1 += x - 1 - i
+        u1 = rankx.sum() - n1*(n1+1)/2
+        u2 = n1 * n2 - u1
+        if alternative == 'two-sided':
+            smallu = min(u1, u2)
+        elif alternative == 'greater':
+            smallu = u1
+        else:
+            smallu = u2
+        p = sum(u <= smallu) / len(u)
+        u = smallu
+    else:
+        u1 = n1*n2 + n1*(n1+1)/2.0 - np.sum(rankx, axis=0)  # calc U for x
+        u2 = n1*n2 - u1                            # remainder is U for y
+        if alternative == 'two-sided':
+            bigu = max(u1, u2)
+            smallu = min(u1, u2)
+        elif alternative == 'greater':
+            bigu = u2
+            smallu = u1
+        else:
+            bigu = u1
+            smallu = u2
```

Comment:
This bit, starting from line 4407 (`u1 = ...`) is duplicated between the two `if exact` branches, up to what seems to be swapping around u1 and u2. I suggest deduplicating.